### PR TITLE
 ASoC: Intel: skl_hda_dsp_generic: add dmic support on hda platforms

### PR DESCRIPTION
**Pull request by RanderWang**
Now two dmics are supported on skl+ platforms. Remove codec check
, report all supported dai links to ASoC and topolog would choose
supported dai links

Tested on WHL & ICL

Signed-off-by: Rander Wang <rander.wang@linux.intel.com>

### CODE REVIEW

**Comment by RanderWang** [comment]
We don't need to add code check whether this is only HDMI or not, just do as other machine driver in intel/boards. This check makes thing complex for cases HDMI, HDMI+DMIC, HDMI+HDAC, HDMI+HDAC+DMIC.  Just report all the supported dai links and let topology choose them.

File: sound/soc/intel/boards/skl_hda_dsp_generic.c [hunk 1]
```diff
@@ -55,6 +55,9 @@ static const struct snd_soc_dapm_route skl_hda_map[] = {
 	{ "Digital CPU Capture", NULL, "Digital Codec Capture" },
 	{ "codec2_in", NULL, "Alt Analog CPU Capture" },
 	{ "Alt Analog CPU Capture", NULL, "Alt Analog Codec Capture" },
+
+	{"dmic01", NULL, "DMIC AIF"},
+	{"dmic02", NULL, "DMIC AIF"},
```
Comment (bardliao) [inline sound/soc/intel/boards/skl_hda_dsp_generic.c]: Do we need this? It seems there is no dmic widget in the topology.

File: sound/soc/intel/boards/skl_hda_dsp_common.c [hunk 2]
```diff
@@ -93,6 +93,28 @@ struct snd_soc_dai_link skl_hda_be_dai_links[HDA_DSP_MAX_BE_DAI_LINKS] = {
...
+		.name = "dmic02",
+		.id = 7,
+		.cpu_dai_name = "DMIC16k Pin",
+		.codec_name = "dmic-codec",
+		.codec_dai_name = "dmic-hifi",
+		.platform_name = "0000:00:1f.3",
+		.dpcm_capture = 1,
+		.init = NULL,
+		.no_pcm = 1,
+	},
```
Comment (plbossart) [inline sound/soc/intel/boards/skl_hda_dsp_common.c:113]: Can you test this with the SKL driver as well and show this doesn't harm existing solutions

File: sound/soc/intel/boards/skl_hda_dsp_generic.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/boards/skl_hda_dsp_generic.c]: same, not sure if this is needed for SOF, and it's not supported by the SKL/cAVS driver.

File: sound/soc/intel/boards/skl_hda_dsp_generic.c [hunk 3]
```diff
@@ -96,35 +99,16 @@ static struct snd_soc_card hda_soc_card = {
...
-		return -EINVAL;
-	}
+	int i;
 
-	card->num_links = num_links;
-	card->num_dapm_routes = num_route;
+	card->num_links = ARRAY_SIZE(skl_hda_be_dai_links);
+	card->num_dapm_routes = ARRAY_SIZE(skl_hda_map),
+	card->num_dapm_widgets = ARRAY_SIZE(skl_hda_widgets);
+	card->dapm_widgets = skl_hda_widgets;
```
Comment (plbossart) [inline sound/soc/intel/boards/skl_hda_dsp_generic.c]: you are changing the functionality here. If there is no hdaudio codec, then we only want the idisp dailinks. And you need to keep testing for the codec count, we can only do 1 or 2 codecs.

Comment (plbossart) [reply sound/soc/intel/boards/skl_hda_dsp_generic.c]: please test this on Up2 with HDMI+ DMIC.

File: sound/soc/intel/boards/skl_hda_dsp_generic.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/intel/boards/skl_hda_dsp_generic.c]: Actually, this is the widget in dmic codec driver, so we need to connect to it, just like "Alt Analog CPU Capture" which is defined in hdac_hda codec.
````
static const struct snd_soc_dapm_route intercon[] = {
	{"DMIC AIF", NULL, "DMic"},
};

````

Comment (bardliao) [reply sound/soc/intel/boards/skl_hda_dsp_generic.c]: @RanderWang Yes, "DMIC AIF" is from dmic codec driver, but where is "dmic01" and "dmic02"?

File: sound/soc/intel/boards/skl_hda_dsp_generic.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/intel/boards/skl_hda_dsp_generic.c]: yes, I misunderstood the algorithm in ASOC. For Up2, A error of no codec dai link would be reported.

File: sound/soc/intel/boards/skl_hda_dsp_generic.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/intel/boards/skl_hda_dsp_generic.c]: dmic01 and dmic02 is created by topology

**Comment by RanderWang** [comment]
Refine my patch
(1) check dmic_num to set dai link
(2) move dmic dapm route to topology because the dapm route table is constant and I can't change it dynamically

Tested it on WHL & UP2 & cAVS driver on Gemilake.
Thanks!

File: sound/soc/intel/boards/skl_hda_dsp_generic.c [hunk 4]
```diff
@@ -114,9 +117,27 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
 	if (codec_count == 1 && codec_mask & IDISP_CODEC_MASK) {
 		num_links = IDISP_DAI_COUNT;
 		num_route = IDISP_ROUTE_COUNT;
+
+		if (mach_params->dmic_num) {
```
Comment (plbossart) [inline sound/soc/intel/boards/skl_hda_dsp_generic.c]: @RanderWang how does this work? there is no dmic_num in mach_params?

Comment (plbossart) [reply sound/soc/intel/boards/skl_hda_dsp_generic.c]: never mind, I didn't even remember I added this :-(

File: sound/soc/sof/intel/cnl.c [hunk 5]
```diff
@@ -252,6 +252,7 @@ const struct sof_intel_dsp_desc cnl_chip_info = {
 	.ipc_ack = CNL_DSP_REG_HIPCIDA,
 	.ipc_ack_mask = CNL_DSP_REG_HIPCIDA_DONE,
 	.ipc_ctl = CNL_DSP_REG_HIPCCTL,
-	.rom_init_timeout	= 300,
+	.rom_init_timeout = 300,
+	.dmic_num = 2,
```
Comment (plbossart) [inline sound/soc/sof/intel/cnl.c]: what does this represent? the number of PDM interfaces, the number of dailinks, the number of actual microphones? The latter can not possibly come from fixed descriptors but would need to come from topology or NHLT.

Comment (RanderWang) [reply sound/soc/sof/intel/cnl.c]: It is the number of dai links for dmic supported by SOF driver and cAVS driver.
````
They are defined in 
struct snd_soc_dai_driver skl_dai[] or static struct snd_soc_dai_driver skl_platform_dai[]
= {
...
{
	.name = "DMIC01 Pin",
},
{
	.name = "DMIC16k Pin",
},
````

**Comment by plbossart** [comment]
> +	.dmic_num = 2,
> 
> It is the number of dai links for dmic supported by SOF driver and cAVS 
> driver.
> 
> |They are defined in struct snd_soc_dai_driver skl_dai[] or static 
> struct snd_soc_dai_driver skl_platform_dai[] = { ... { .name = "DMIC01 
> Pin", }, { .name = "DMIC16k Pin", }, |

Well then we have a problem
in other drivers this is the actual number of microphones, as extracted 
from NHLT, so you are using the mach_params->dmic_num in different ways.

**Comment by RanderWang** [comment]
> Well then we have a problem in other drivers this is the actual number of microphones, as extracted from NHLT, so you are using the mach_params->dmic_num in different ways.

Yes, it is a diversity of usage. Actually, dmic_num is used as dmic channel number in cAVS dsp driver and machine driver.
````
in dsp driver, dmic_num = dmic_geo;

			case NHLT_MIC_ARRAY_2CH_BIG:
				dmic_geo |= MIC_ARRAY_2CH;
				break;
                        ...
			case NHLT_MIC_ARRAY_4CH_2ND_GEOM:
				dmic_geo |= MIC_ARRAY_4CH;
				break;

in I2S sklylake machine driver:
		dmic_constraints = mach->mach_params.dmic_num == 2 ?
			&constraints_dmic_2ch : &constraints_dmic_channels;
````
And the dmic dai link is constant in skylake machine drivers and never changed. There is only one dmic dai link in all the machine drivers except one: bxt_rt298 which creates two dmic dai links statically. 

We could also always report two dmic dai links to ASoC and ignore dmic_num in machine driver. I tested on cAVS driver on leafhill and found there is no dmic dai link in the available topology file for cAVS driver, So I can't test capture with dmic. The dmic_num (dmic channel number) is gotten from BIOS and for snd_pcm_hw_constraint in cAVS driver which fix rate to 48k and channel number to dmic_number. But we don't do it now and it doesn't work for uboot or coreboot. So do we need to apply dmic_num to hw constarint in machine driver ?

I will update my PR for reference.
Thanks!

**Comment by plbossart** [comment]
@RanderWang I will have to put your PR on the back-burner until next week, too many questions on what the definition of those values and how to be consistent. I really don't think we can put hard-coded mic numbers in the platform side, it's got to come from topology and/or acpi.

**Comment by RanderWang** [comment]
Thanks.   I talked with Mengdong and Bard, got one idea: the dai links in topology is a subset of dai links in machine driver. Some dai links is not supported by topology but they are reported by machine driver. So I think it is not good to get dai links from topology for machine driver. And also it is impossible for topology is loaded after machine driver which trigger Asoc to load topology.

**Comment by plbossart** [comment]
> Thanks. I talked with Mengdong and Bard, got one idea: the dai links in topology is a subset of dai links in machine driver. Some dai links is not supported by topology but they are reported by machine driver. So I think it is not good to get dai links from topology for machine driver. And also it is impossible for topology is loaded after machine driver which trigger Asoc to load topology.

So what's the idea and the solution then? what are you suggesting?

**Comment by RanderWang** [comment]
> loaded after machine driver which trigger Asoc to load topology.
>
> So what's the idea and the solution then? what are you suggesting?

I also discussed with keyon.  Two solution: (1) the second version of this PR, report all the dmic dia links to ASoC, let topology choose them. (2) the first version of this PR, check dmic_num to report dmic dai links. The key issue is how to get dmic_num. We can't get it from topology and BIOS, so we have to query it from each platforms info table.  But dmic_num is the number of dmic channel in cAVS driver, so it should be configured dynamically. So I don't like the solution (2).

**Comment by plbossart** [comment]
> > loaded after machine driver which trigger Asoc to load topology.
> > So what's the idea and the solution then? what are you suggesting?
> 
> I also discussed with keyon. Two solution: (1) the second version of this PR, report all the dmic dia links to ASoC, let topology choose them. (2) the first version of this PR, check dmic_num to report dmic dai links. The key issue is how to get dmic_num. We can't get it from topology and BIOS, so we have to query it from each platforms info table. But dmic_num is the number of dmic channel in cAVS driver, so it should be configured dynamically. So I don't like the solution (2).

Humm... I read a set of sentences that don't seem to have a logical connection, so I have no idea what you are suggesting. your preferred solution (1) does not address the fact the dmic_num argument passed to the machine driver is not handled by topology, so where does this leave us?

**Comment by RanderWang** [comment]
> Humm... I read a set of sentences that don't seem to have a logical connection, so I have no idea what you are suggesting. your preferred solution (1) does not address the fact the dmic_num argument passed to the machine driver is not handled by topology, so where does this leave us?

There is no error reported by ASoC if there is no FE using the dai links.,  And both SOF and cAVS driver
have registered dmic device which would register dmic codec dai, so there is no error for no codec dai found. I tested this case on SOF and cAVS driver.

**Comment by plbossart** [comment]
> > Humm... I read a set of sentences that don't seem to have a logical connection, so I have no idea what you are suggesting. your preferred solution (1) does not address the fact the dmic_num argument passed to the machine driver is not handled by topology, so where does this leave us?
> 
> There is no error reported by ASoC if there is no FE using the dai links., And both SOF and cAVS driver
> have registered dmic device which would register dmic codec dai, so there is no error for no codec dai found. I tested this case on SOF and cAVS driver.

Sorry @RanderWang, I have no idea what you are referring to. I realize English is a complicated language but at this point I don't know what problem you are trying to fix nor what the recommended solution is.

**Comment by wenqingfu** [comment]
> Now two dmics are supported on skl+ platforms.

Frankly, you can have more than two if you talk with @singalsu . We need a more generic approach.

**Comment by RanderWang** [comment]
> > Now two dmics are supported on skl+ platforms.
> 
> Frankly, you can have more than two if you talk with @singalsu . We need a more generic approach.

@wenqingfu   This PR is for dmic dai links, not for DMIC number.  I have explained that our SOF driver and cAVS driver only support two dmic dai links, so why add more than that ?

**Comment by RanderWang** [comment]
> > > Humm... I read a set of sentences that don't seem to have a logical connection, so I have no idea what you are suggesting. your preferred solution (1) does not address the fact the dmic_num argument passed to the machine driver is not handled by topology, so where does this leave us?
> > 
> > 
> > There is no error reported by ASoC if there is no FE using the dai links., And both SOF and cAVS driver
> > have registered dmic device which would register dmic codec dai, so there is no error for no codec dai found. I tested this case on SOF and cAVS driver.
> 
> Sorry @RanderWang, I have no idea what you are referring to. I realize English is a complicated language but at this point I don't know what problem you are trying to fix nor what the recommended solution is.

It is not only caused by English language, but also the thinking pattern. I would speak out my suggestion : I prefer this PR now.  what I explain is that: dmic_num can't be used by SOF, and this PR works on both SOF and cAVS driver and why.  And what is your idea ?

**Comment by wenqingfu** [comment]
> Thanks. I talked with Mengdong and Bard, got one idea: the dai links in topology is a subset of dai links in machine driver. Some dai links is not supported by topology but they are reported by machine driver. So I think it is not good to get dai links from topology for machine driver. And also it is impossible for topology is loaded after machine driver which trigger Asoc to load topology.

Can you add the support through topology? If so, it looks cleaner as topology is supposed to be an ASoC feature to provide such information. If not, can you please rephrase the above paragraph to explain why? I see you tried but I can hardly get your point, either.

**Comment by RanderWang** [comment]
sorry, I really didn't get the point. I tried to prove my PR work.    Topology only defines front-end dai links, and machine driver defines back-end dai links, so we can't set back-end dmic dai link in topology. And there is no support in ASoC for back-end dai link defined by topology.  Thanks to @mengdonglin , she provided the background:  https://mailman.alsa-project.org/pipermail/alsa-devel/2015-August/096451.html.

````
I'm really confused as to why physical links that exist in the board are
coming in via topology information loaded from userspace - it's not like
the DSP firmware is going to be able to physically connect the DSP to
things where no links existed before.  I'd expect that these links are
connections between physical devices which just exist in the system and
that while the topology information might leave some of them dangling
they wouldn't be created or removed.

Presumably we're also going to need to handle the case where the
firmware is unloaded and we destroy the links if we do this...
````

**Comment by plbossart** [comment]
This PR has reached a state where additional dialogue looks like a waste of time. There is no way I will merge this code as is until there is a clarification on how the number of dmics are handled.

**Comment by RanderWang** [comment]
Thanks Pierre show me the direction.  Now I have a solution and a question.

I can get the capability of dmic link count from a register and set it in dmic_num checked by machine driver.  I don't know whether the dmic link count (I checked the register and got 2) is the number of dmics you commended.

And the dmic_num in snd_soc_acpi_mach_params is for channel number of dmic array. It was first introduced by https://mailman.alsa-project.org/pipermail/alsa-devel/2016-May/108207.html. I don't find a way to get this information without reading nhlt and  the dmic dai links number would be defined by this dmic_num?

Thanks!

**Comment by singalsu** [comment]
I'm not sure what's the open but most Intel platforms support two DMIC DAIs. There will be also some with one. The FW DMIC driver currently uses only platform macros to set up the DAI count. Eventually it should read the SHIM configuration bits to determine it in run-time. 

Also the max. number of stereo mic pairs supported by platform is today controlled by platform macros but that information is as more reliable found from too SHIM, should change. So the FW could provide the DMIC configuration information via IPC or the driver could read it directly.

The NHLT information mic count is similar to what SOF topology provides for FW  driver (microphones count & connection). We don't have array geometry IPC since it's for higher functionality level (beamforming).

The NHTL information as seen in the ALSA  patch looks quite coarse for mic count. It reports how many microphones there are but not to which stereo PDM controllers they hook. If the devices would populate the microphones sequentially starting from channel 1 it would work but there are devices with non-connected channels in-between. It could be used to replace part of topology configuration for DMIC but maybe there should be a way to override if the mic connections are not as assumed.

There's also much lower level DMIC configuration information in NHLT but we don't use it at the moment in SOF. It could be used none of it (today) / partial(determine just connected mic) / full. The lowest level information in NHLT is very platform dependent.

Finally there are also seen some requirements to set up and capture max. amount of channels in the platform and let user space to control channels select mux to extract connected or use case specific microphones.

I don't know if this cleared the issue but wanted to add just my thoughts on the challenges and help you decide where/how to handle the DAI and mic count/channels.

**Comment by plbossart** [comment]
> I'm not sure what's the open but most Intel platforms support two DMIC DAIs. There will be also some with one. The FW DMIC driver currently uses only platform macros to set up the DAI count. Eventually it should read the SHIM configuration bits to determine it in run-time.

@singalsu The current definitions in machine drivers is one DAI for the first level of decimation (48k) and one for the second level (16k). If we can read the number of DAIs from the SHIM registers then we have a definition problem between kernel and firmware.
Can you elaborate on what you view as a DAI?

**Comment by singalsu** [comment]
@plbossart  I meant that DAI equals a HW FIFO. Currently all platforms supported by SOF have two DMIC FIFOs.

**Comment by plbossart** [comment]
> @plbossart I meant that DAI equals a HW FIFO. Currently all platforms supported by SOF have two DMIC FIFOs.

well then clearly we have a disconnect with the way the cAVS driver works and defined dais/dailinks. I am not saying it's necessarily bad but it'll require changes to all machine drivers making use of DMICs.

**Comment by plbossart** [comment]
> > @plbossart I meant that DAI equals a HW FIFO. Currently all platforms supported by SOF have two DMIC FIFOs.
> 
> well then clearly we have a disconnect with the way the cAVS driver works and defined dais/dailinks. I am not saying it's necessarily bad but it'll require changes to all machine drivers making use of DMICs.

Also how do you count FIFOs? Is there one FIFO for 16kHz and one for 48kHz, times 2 for 2 interfaces?

**Comment by singalsu** [comment]
The DMIC FIFOs are "global" so there's only one primary (48k) and one secondary (16k) FIFO. The FIFOs are multi-channel, e.g 4ch or 8ch depending on number of stereo PDM controllers supported by the platform.

We have one DMIC instance so the times 2 scenario from your reply doesn't exist. Not yet at least.

**Comment by plbossart** [comment]
> The DMIC FIFOs are "global" so there's only one primary (48k) and one secondary (16k) FIFO. The FIFOs are multi-channel, e.g 4ch or 8ch depending on number of stereo PDM controllers supported by the platform.
> 
> We have one DMIC instance so the times 2 scenario from your reply doesn't exist. Not yet at least.

Sorry, I am lost here.

Each PDM interface has 2 wires and can deal with 2channels max, so there must be a level of aggregation I am missing. Is that level in hardware (controller and/or DMA) or at the firmware level?

**Comment by ranj063** [comment]
@plbossart Sorry to divert from the last conversation but coming back to the original intent of this PR, this is what I can summarize:
1. Intention: Add DMIC capture pipelines to the hda-generic-idsp topology
2. Solution: For cavs platforms, the number of DMIC FIFO's is only 2. So it should be OK to hard code it to 2 in the machine driver and add the DMIC dai links.  

Given the above 2, I get the feeling that this PR is reasonable. What do you think?

File: sound/soc/intel/boards/skl_hda_dsp_common.c [hunk 6]
```diff
@@ -93,6 +93,28 @@ struct snd_soc_dai_link skl_hda_be_dai_links[HDA_DSP_MAX_BE_DAI_LINKS] = {
...
 	},
+	{
+		.name = "dmic01",
+		.id = 6,
+		.cpu_dai_name = "DMIC01 Pin",
+		.codec_name = "dmic-codec",
+		.codec_dai_name = "dmic-hifi",
+		.platform_name = "0000:00:1f.3",
+		.dpcm_capture = 1,
+		.init = NULL,
```
Comment (plbossart) [inline sound/soc/intel/boards/skl_hda_dsp_common.c]: init doesn't need to be set.

File: sound/soc/intel/boards/skl_hda_dsp_common.c [hunk 7]
```diff
@@ -93,6 +93,28 @@ struct snd_soc_dai_link skl_hda_be_dai_links[HDA_DSP_MAX_BE_DAI_LINKS] = {
...
+	},
+	{
+		.name = "dmic16k",
+		.id = 7,
+		.cpu_dai_name = "DMIC16k Pin",
+		.codec_name = "dmic-codec",
+		.codec_dai_name = "dmic-hifi",
+		.platform_name = "0000:00:1f.3",
+		.dpcm_capture = 1,
+		.init = NULL,
```
Comment (plbossart) [inline sound/soc/intel/boards/skl_hda_dsp_common.c]: same, no need to set this.

**Comment by singalsu** [comment]
@plbossart Yep, the DMIC channels aggregation happens at the common HW logic. The FIFOs contain channels interleaved data that's directly compatible with multichannel S16_LE and S32_LE format of SOF. The channels interleaving/packing is HW registers controllable.

E.g. the primary FIFO (DAI[0]) has a mux at input so the data can be written as interleaved from the FIR A outputs of each stereo decimator. Similarly the data into secondary FIFO is written via a mux from FIR B outputs. Each stereo decimator contains the A and B output.

**Comment by RanderWang** [comment]
> Still I would like to challenge the tests done on GLK with the Skylake driver. GLK does not have an HDaudio codec, so how is it relevant? 

I tested on a NUC with Gemilake and there is a HDaudio codec. But the topology I used is gotten from a LeafHill and there is no dmic in this topology. So I can't validate dmic, just make sure my PR produce no error. 

> Also you'd need to specify which topology is used. And last all existing machine driver have a _dmic_fixup routine, and it's not clear to me if this would be needed.

The dmic_fixup function all set channel->min & max to 2 or 4 according to params.  It means user
could only use 2 or 4 channel and this make life easy for driver. I will add this in next update.

> And one more for the road: I still don't get if the dmic_num parameter is needed or not.

 some machine driver use dmic_num, which is set in BIOS like 2ch Dmic Array,  to constraint hw channel number like dmic_fixup. I checked all existing skylake machine drivers and found some of them don't use it, eg skl_rt286.c. And these machine drivers,  which use dmic_num, are bespoken for special device. SOF don't get dmic_num from BIOS, so it is not fit for us. My idea is to do hw constraint in dmic_fixup.

Thanks!

File: sound/soc/intel/boards/skl_hda_dsp_common.c (hunk 7 above)
Comment (RanderWang) [reply sound/soc/intel/boards/skl_hda_dsp_common.c]: I get it, thanks

File: sound/soc/intel/boards/skl_hda_dsp_common.c (hunk 6 above)
Comment (RanderWang) [reply sound/soc/intel/boards/skl_hda_dsp_common.c]: I get it, thanks

**Comment by RanderWang** [comment]
Update my PR, tested it on whl and GLK with skylake driver.

one finding:  we don't need to add dmic_fixup function in this machine driver which should work with topology.  For existing intel machine driver,  we do in skl_rt286.c like : 
````
	{
		.name = "dmic01",
		.id = 1,
		.cpu_dai_name = "DMIC01 Pin",
		.codec_name = "dmic-codec",
		.codec_dai_name = "dmic-hifi",
		.platform_name = "0000:00:1f.3",
		.be_hw_params_fixup = skylake_dmic_fixup,
		.ignore_suspend = 1,
		.dpcm_capture = 1,
		.no_pcm = 1,
	},
````
Actually,  this could only work for legacy machine driver which doesn't work with topology.
If the machine driver works with topology, the behavior is changed by this patch:  https://github.com/thesofproject/linux/commit/a655de808cbde6c58b3298e704d786b53f59fb5d

In this patch fixup function would be overwriten:
````
			/* override any BE fixups */
			dai_link->be_hw_params_fixup =
				component->driver->be_hw_params_fixup;
````

For SOF, the fixup function is sof_pcm_dai_link_fixup and seppo has done some job. And we don't need to restrict channel number which is defined in topology. If the channel number is not correct, there will be error:
````
arecord: set_params:1305: Channels count non available
````
There is no such fixup function in skylake driver. And there is no need to restrict channel number according to above reason.

Thanks

File: sound/soc/intel/boards/skl_hda_dsp_generic.c [hunk 8]
```diff
@@ -112,11 +115,23 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
...
+		 */
+		for (i = 0; i < DMIC_DAI_COUNT; i++) {
+			skl_hda_be_dai_links[IDISP_DAI_COUNT + i] =
+				skl_hda_be_dai_links[IDISP_DAI_COUNT +
+					HDAC_DAI_COUNT + i];
+		}
 	} else if (codec_count == 2 && codec_mask & IDISP_CODEC_MASK) {
 		num_links = ARRAY_SIZE(skl_hda_be_dai_links);
-		num_route = ARRAY_SIZE(skl_hda_map),
+		num_route = ARRAY_SIZE(skl_hda_map);
```
Comment (plbossart) [inline sound/soc/intel/boards/skl_hda_dsp_generic.c:134]: How did this work in the first place with the comma at the end? 
Please fix this in a separate patch and add a Fixes: 6bae5ea949892 tag to your commit, this needs to go upstream.

Comment (RanderWang) [reply sound/soc/intel/boards/skl_hda_dsp_generic.c:134]: Actually, comma also works and it is supported by c programming syntax, So the new patch is not a fix, just a refinement.

**Comment by RanderWang** [comment]
Update my comments,  thanks!

**Pull request merged**